formulaic/materializers/transforms/basis_splines.py

```diff
@@ -0,0 +1,86 @@
+
+    Parameters
+    --------------
+    data:
+    n_knots : integer, optional
+        Number of knots requested. Options for knots include any positive integer if the location of knots are
+        specified. If knot locations are not specified, n_knots must be an integer between 1 to 7. Default is 3 knots
+    knots : list, optional
+        Location of specified knots in a list. To specify the location of knots, put desired numbers for knots into a
+        list. Be sure that the length of the list is the same as the specified number of knots. Default is None, so
+        that the function will automatically determine knot locations without user specification
+    term : integer, float, optional
+        High order term for the spline terms. To calculate a quadratic spline change to 2, cubic spline
+        change to 3, etc. Default is 1, i.e. a linear spline
+    restricted : bool, optional
+        Whether to return a restricted spline. Note that the restricted spline returns one less column than the number
+        of knots. An unrestricted spline returns the same number of columns as the number of knots. Default is False,
+        providing an unrestricted spline
+
+    """
+    if knots is None:
+        if n_knots == 1:
+            knots = [50]
+        elif n_knots == 2:
+            knots = [100 / 3, 200 / 3]
+        elif n_knots == 3:
+            knots = [5, 50, 95]
+        elif n_knots == 4:
+            knots = [5, 35, 65, 95]
+        elif n_knots == 5:
+            knots = [5, 27.5, 50, 72.5, 95]
+        elif n_knots == 6:
+            knots = [5, 23, 41, 59, 77, 95]
+        elif n_knots == 7:
+            knots = [2.5, 1100 / 60, 2600 / 75, 50, 7900 / 120, 4900 / 60, 97.5]
+        else:
+            raise ValueError(
+                "When the knot locations are not pre-specified, the number of specified knots must be"
+                " an integer between 1 and 7")
+        pts = np.percentile(data, q=knots).tolist()
+    else:
+        if n_knots != len(knots):
+            raise ValueError("The number of knots and the number of specified knots must match")
+        pts = sorted(knots)
+
+    def bs_transform(x):
+        columns = np.arange(n_knots)
+        x = np.asarray(x)
+        V = np.empty((x.shape[0], len(pts)))
+
+        for i, pt in enumerate(pts):
+            V[:, i] = np.where(x > pt, (x - pt) ** term, 0)
+            V[:, i] = np.where(pd.isnull(x), np.nan, V[:, i])
+        if restricted is False:
+            return pd.DataFrame(V, columns=columns)
+
+        else:
+            for i, pt in enumerate(pts):
+                V[:, i] = np.where(x > pt, V[:, i] - V[:, -1], 0)
+                V[:, i] = np.where(pd.isnull(x), np.nan, V[:, i])
+            return pd.DataFrame(V[:, :-1], columns=columns[:-1])
+
+    if 'bs_transform' not in state:
+        state['bs_transform'] = bs_transform
+
+    encoded = dict(state['bs_transform'](data))
+    encoded.update({
+        '__format__': "{name}[{field}]",
```

Comment:
other metadata that I saw was `__kind__` and `__encoded__` - not sure if those are relevant?

Reply:
Aye... sorry for the state of documentation. It's a fledgling project, but it _will_ get better!

`__encoded__` tells the formulaic materializer that the data has already been encoded and doesn't need to go through any default encoding routines.
`__kind__` tells the materializer what kind of data is stored in the output of your transform (categorical, etc); which allows the materializers to know whether it needs to worry about structural full-rankness / etc.

In this case, I think omitting both should be fine (since the result is numeric vectors, which have pretty innocuous default encoding strategies).